core/templates/services/playthrough.service.ts

```diff
@@ -16,330 +16,350 @@
-  'PlaythroughObjectFactory', 'StopwatchObjectFactory',
-  'UrlInterpolationService', 'ACTION_TYPE_ANSWER_SUBMIT',
-  'ACTION_TYPE_EXPLORATION_QUIT', 'ACTION_TYPE_EXPLORATION_START',
-  'CURRENT_ACTION_SCHEMA_VERSION', 'CURRENT_ISSUE_SCHEMA_VERSION',
-  'EARLY_QUIT_THRESHOLD_IN_SECS', 'ISSUE_TYPE_CYCLIC_STATE_TRANSITIONS',
-  'ISSUE_TYPE_EARLY_QUIT', 'ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS',
-  'NUM_INCORRECT_ANSWERS_THRESHOLD', 'NUM_REPEATED_CYCLES_THRESHOLD',
-  'STORE_PLAYTHROUGH_URL',
-  function(
-      $http, ExplorationFeaturesService, LearnerActionObjectFactory,
-      PlaythroughObjectFactory, StopwatchObjectFactory,
-      UrlInterpolationService, ACTION_TYPE_ANSWER_SUBMIT,
-      ACTION_TYPE_EXPLORATION_QUIT, ACTION_TYPE_EXPLORATION_START,
-      CURRENT_ACTION_SCHEMA_VERSION, CURRENT_ISSUE_SCHEMA_VERSION,
-      EARLY_QUIT_THRESHOLD_IN_SECS, ISSUE_TYPE_CYCLIC_STATE_TRANSITIONS,
-      ISSUE_TYPE_EARLY_QUIT, ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS,
-      NUM_INCORRECT_ANSWERS_THRESHOLD, NUM_REPEATED_CYCLES_THRESHOLD,
-      STORE_PLAYTHROUGH_URL) {
-    var playthrough = null;
-    var expStopwatch = null;
-    var isLearnerInSamplePopulation = null;
+@Injectable({
+  providedIn: 'root'
+})
+export class PlaythroughService {
+  constructor(
+    private http: HttpClient,
+    private explorationFeaturesService: ExplorationFeaturesService,
+    private learnerActionObjectFactory: LearnerActionObjectFactory,
+    private playthroughObjectFactory: PlaythroughObjectFactory,
+    private stopwatchObjectFactory: StopwatchObjectFactory,
+    private urlInterpolationService: UrlInterpolationService) {}
 
-    var multipleIncorrectStateName = {
+    playthrough: Playthrough = null;
+    expStopwatch: Stopwatch = null;
+    isLearnerInSamplePopulation: boolean = null;
+    multipleIncorrectStateName = {
       state_name: null,
       num_times_incorrect: null
     };
-
-    var cycleIdentifier = {
+    cycleIdentifier = {
       cycle: null,
       num_cycles: null
     };
-    var visitedStates = [];
-
-    var misTracker = false;
-    var cstTracker = false;
+    visitedStates = [];
+    misTracker: boolean = false;
+    cstTracker: boolean = false;
 
-    var removeOldQuitAction = function() {
-      var quitAction = playthrough.actions[playthrough.actions.length - 1];
+    private _removeOldQuitAction(): void {
+      var quitAction = this.playthrough.actions[
+        this.playthrough.actions.length - 1];
       // After the second quit action is recorded, the first quit is removed
       // using this method. This ensures that there are only two quit actions
       // in the playthrough actions list at a time.
-      playthrough.actions = playthrough.actions.filter(function(action) {
-        return action.actionType !== ACTION_TYPE_EXPLORATION_QUIT;
-      });
-      playthrough.actions.push(quitAction);
-    };
+      this.playthrough.actions = this.playthrough.actions.filter(
+        function(action) {
```

Comment:
Please change `function` to fat arrow () => everywhere.

Reply:
Done.